qiskit/aqua/operators/gradients/circuit_qfis/overlap_block_diag.py

```diff
@@ -0,0 +1,305 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""The module for Quantum the Fisher Information."""
+
+import copy
+from functools import cmp_to_key
+from typing import List, Union, Optional
+
+import numpy as np
+from scipy.linalg import block_diag
+from qiskit.aqua import AquaError
+from qiskit.aqua.operators import ListOp, CircuitOp
+from qiskit.aqua.operators.expectations import PauliExpectation
+from qiskit.aqua.operators.operator_globals import I, Z, Y, X, Zero
+from qiskit.aqua.operators.state_fns import StateFn, CircuitStateFn
+from qiskit.circuit import Parameter, ParameterVector, ParameterExpression
+from qiskit.circuit.library import RZGate, RXGate, RYGate
+from qiskit.converters import dag_to_circuit, circuit_to_dag
+
+from .circuit_qfi import CircuitQFI
+from ..derivative_base import DerivativeBase
+
+
+class OverlapBlockDiag(CircuitQFI):
+    r"""Compute the block-diagonal of the Quantum Fisher Information (QFI) given a pure,
+    parametrized quantum state. The blocks are given by all parameterized gates in quantum circuit
+    layer.
+
+    The QFI is:
+
+        [QFI]kl= Re[〈∂kψ|∂lψ〉−〈∂kψ|ψ〉〈ψ|∂lψ〉] * 4.
+    """
+
+    def convert(self,
+                operator: Union[CircuitOp, CircuitStateFn],
+                params: Optional[Union[ParameterExpression, ParameterVector,
+                                       List[ParameterExpression]]] = None
+                ) -> ListOp:
+
+        r"""
+        Args:
+            operator: The operator corresponding to the quantum state |ψ(ω)〉for which we compute
+                the QFI
+            params: The parameters we are computing the QFI wrt: ω
+
+        Returns:
+            ListOp[ListOp] where the operator at position k,l corresponds to QFI_kl
+
+        Raises:
+            NotImplementedError: If ``operator`` is neither ``CircuitOp`` nor ``CircuitStateFn``.
+
+        """
+        if not isinstance(operator, (CircuitOp, CircuitStateFn)):
```

Comment:
Why does this allow a CircuitOp and CircuitStateFn but LinCombFull only a CircuitStateFn?

Reply:
In principle, the LinCombFull could probably accept a CircuitOp but I'm happy to just restrict both of them to CircuitStateFns for now.

Reply:
Agreed